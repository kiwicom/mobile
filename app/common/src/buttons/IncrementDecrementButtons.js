// @flow

import * as React from 'react';
import { View, StyleSheet } from 'react-native';

import Button from './Button';
import Color from '../Color';

type Props = {|
  onIncrement: () => void,
  onDecrement: () => void,
  number: number,
  min?: number,
  max?: number,
|};

export default function IncrementDecrementButtons(props: Props) {
  const disableDecrement = props.number === props.min;
  const disableIncrement = props.number === props.max;
  return (
    <View style={styles.buttons}>
      <Button
        title="âˆ’" // minus sign, not hyphen
        onPress={!disableDecrement ? props.onDecrement : undefined}
        styles={{
          button: [styles.button, styles.decrementButton],
          buttonText: [
            styles.buttonText,
            disableDecrement && styles.buttonDisabled,
          ],
        }}
      />
      <Button
        title="+"
        onPress={!disableIncrement ? props.onIncrement : undefined}
        styles={{
          button: [styles.button, styles.incrementButton],
          buttonText: [
            styles.buttonText,
            disableIncrement && styles.buttonDisabled,
          ],
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  buttons: {
    flexDirection: 'row',
  },
  button: {
    borderWidth: 1,
    borderColor: Color.brand,
    backgroundColor: '#fff',
    height: 26,
    width: 40,
    borderRadius: 3,
    padding: 2,
    flexGrow: 0,
  },
  buttonText: {
    flex: 1,
    lineHeight: 24,
    color: Color.brand,
    fontSize: 28,
    fontWeight: '300',
  },
  incrementButton: {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },
  decrementButton: {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0,
    borderRightWidth: 0,
  },
  buttonDisabled: {
    opacity: 0.55,
  },
});
