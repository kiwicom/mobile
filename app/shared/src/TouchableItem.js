// @flow

import * as React from 'react';
import {
  Platform,
  TouchableNativeFeedback,
  TouchableOpacity,
  View,
} from 'react-native';

import { type StylePropType } from '../types/Styles';

type Props = {
  // not exact: additional props allowed for TouchableOpacity and TouchableNativeFeedback
  children: React.Node,
  borderless?: boolean,
  pressColor?: string,
  style?: StylePropType,
};

const ANDROID_VERSION_LOLLIPOP = 21;

/**
 * TouchableItem renders a touchable that looks native on both iOS and Android.
 *
 * It provides an abstraction on top of TouchableNativeFeedback and
 * TouchableOpacity.
 *
 * On iOS you can pass the props of TouchableOpacity, on Android pass the props
 * of TouchableNativeFeedback.
 *
 * Note: this is basically copy-paste from React Navigation
 *
 * @see https://github.com/react-navigation/react-navigation/blob/cf36f22e68fe4a7f475d714590f665dfc29e3e25/src/views/TouchableItem.js
 */
export default class TouchableItem extends React.Component<Props> {
  static defaultProps = {
    borderless: false,
    pressColor: 'rgba(0, 0, 0, .32)',
  };

  render = () => {
    /*
     * TouchableNativeFeedback.Ripple causes a crash on old Android versions,
     * therefore only enable it on Android Lollipop and above.
     *
     * All touchables on Android should have the ripple effect according to
     * platform design guidelines.
     * We need to pass the background prop to specify a borderless ripple effect.
     */
    if (
      Platform.OS === 'android' &&
      Platform.Version >= ANDROID_VERSION_LOLLIPOP
    ) {
      const { style, ...rest } = this.props;
      return (
        <TouchableNativeFeedback
          useForeground={TouchableNativeFeedback.canUseNativeForeground()}
          {...rest}
          style={null}
          background={TouchableNativeFeedback.Ripple(
            this.props.pressColor,
            this.props.borderless,
          )}
        >
          <View style={style}>{React.Children.only(this.props.children)}</View>
        </TouchableNativeFeedback>
      );
    }

    // React.Children.only is only necessary for TouchableNativeFeedback
    // but since we are building for both platforms it is here as well
    // to discover this mistake as soon as possible
    return (
      <TouchableOpacity {...this.props}>
        {React.Children.only(this.props.children)}
      </TouchableOpacity>
    );
  };
}
