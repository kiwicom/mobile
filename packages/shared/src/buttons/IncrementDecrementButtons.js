// @flow srict

import * as React from 'react';
import { View } from 'react-native';
import { Translation } from '@kiwicom/mobile-localization';
import { defaultTokens } from '@kiwicom/mobile-orbit';

import Text from '../Text';
import TextIcon from '../icons/TextIcon';
import Touchable from '../Touchable';
import StyleSheet from '../PlatformStyleSheet';
import type { StylePropType } from '../../types/Styles';

type ButtonProps = {|
  +icon: React.Element<typeof TextIcon>,
  +touchable: boolean,
  +onPress: () => void,
  +style?: StylePropType,
|};

const Button = ({ icon, touchable, onPress }: ButtonProps) => {
  const inner = (
    <View style={[styles.button, touchable ? null : styles.buttonDisabled]}>
      {icon}
    </View>
  );

  if (touchable) {
    return <Touchable onPress={onPress}>{inner}</Touchable>;
  }

  return inner;
};

type Props = {|
  +onIncrement: () => void,
  +onDecrement: () => void,
  +number: number,
  +showNumber: boolean,
  +numberStyle?: StylePropType,
  +min?: number,
  +max?: number,
|};

export default function IncrementDecrementButtons(props: Props) {
  const disableDecrement = props.number === props.min;
  const disableIncrement = props.number === props.max;

  return (
    <View style={styles.row}>
      <Button
        icon={<TextIcon code="&#xe118;" orbit={true} style={styles.icon} />}
        touchable={!disableDecrement}
        onPress={props.onDecrement}
      />
      {props.showNumber && (
        <Text style={props.numberStyle}>
          <Translation passThrough={props.number} />
        </Text>
      )}
      <Button
        icon={<TextIcon code="&#xe122;" orbit={true} style={styles.icon} />}
        touchable={!disableIncrement}
        onPress={props.onIncrement}
      />
    </View>
  );
}

IncrementDecrementButtons.defaultProps = {
  showNumber: false,
};

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    color: defaultTokens.colorIconSecondary,
    fontSize: 14,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    width: 28,
    height: 28,
    borderRadius: 6,
    backgroundColor: defaultTokens.paletteCloudNormal,
  },
  buttonDisabled: {
    opacity: parseFloat(defaultTokens.opacityButtonDisabled),
  },
});
