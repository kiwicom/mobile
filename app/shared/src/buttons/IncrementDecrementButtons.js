// @flow

import * as React from 'react';
import { View, Text } from 'react-native';

import Color from '../Color';
import TouchableItem from '../TouchableItem';
import StyleSheet from '../PlatformStyleSheet';

const Button = ({
  text,
  touchable,
  onPress,
}: {|
  text: string,
  touchable: boolean,
  onPress: () => void,
|}) => {
  const inner = (
    <View style={styleSheet.button}>
      <Text
        textAnchor="middle"
        style={[
          styleSheet.buttonText,
          {
            opacity: touchable ? 1 : 0.5,
          },
        ]}
      >
        {text}
      </Text>
    </View>
  );

  if (touchable) {
    return (
      <TouchableItem activeOpacity={0.6} onPress={onPress}>
        {inner}
      </TouchableItem>
    );
  }

  return inner;
};

export default function IncrementDecrementButtons(props: {|
  onIncrement: () => void,
  onDecrement: () => void,
  number: number,
  min?: number,
  max?: number,
|}) {
  const disableDecrement = props.number === props.min;
  const disableIncrement = props.number === props.max;

  return (
    <View style={styleSheet.buttonsGroup}>
      <View style={styleSheet.buttonLeft}>
        <Button
          text="â€“"
          touchable={!disableDecrement}
          onPress={props.onDecrement}
        />
      </View>
      <Button
        text="+"
        touchable={!disableIncrement}
        onPress={props.onIncrement}
      />
    </View>
  );
}

const styleSheet = StyleSheet.create({
  buttonsGroup: {
    flexDirection: 'row',
    borderWidth: 1,
    borderColor: Color.brand,
    height: 29,
    width: 94, // 47 * 2 (see button)
    borderRadius: 4,
    backgroundColor: Color.white,
  },
  button: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: 47, // 94 / 2 (see buttonsGroup)
  },
  buttonLeft: {
    alignItems: 'center',
    borderRightWidth: 1,
    borderColor: Color.brand,
  },
  buttonText: {
    color: Color.brand,
    fontSize: 25,
    // this helps to vertically align "+" and "-" on both platforms
    android: {
      lineHeight: 16,
    },
    ios: {
      lineHeight: 25,
    },
  },
});
