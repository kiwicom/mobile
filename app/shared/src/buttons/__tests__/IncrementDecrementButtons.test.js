// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';

import IncrementDecrementButtons from '../IncrementDecrementButtons';
import TouchableItem from '../../TouchableItem';

const renderButtons = (
  onIncrement: Function,
  onDecrement: Function,
  number: number,
) => {
  return renderer.create(
    <IncrementDecrementButtons
      onIncrement={onIncrement}
      onDecrement={onDecrement}
      number={number}
      min={1}
      max={3}
    />,
  );
};

describe('IncrementDecrementButtons', () => {
  it('should increment', () => {
    const onIncrement = jest.fn();
    const onDecrement = jest.fn();
    const testRenderer = renderButtons(onIncrement, onDecrement, 2);
    const testInstance = testRenderer.root;

    testInstance.findAllByType(TouchableItem)[1].props.onPress();

    expect(onIncrement).toBeCalled();
    expect(onDecrement).not.toBeCalled();
  });

  it('should decrement', () => {
    const onIncrement = jest.fn();
    const onDecrement = jest.fn();
    const testRenderer = renderButtons(onIncrement, onDecrement, 2);
    const testInstance = testRenderer.root;

    testInstance.findAllByType(TouchableItem)[0].props.onPress();

    expect(onDecrement).toBeCalled();
    expect(onIncrement).not.toBeCalled();
  });

  it('should not decrement under min', () => {
    const onIncrement = jest.fn();
    const onDecrement = jest.fn();
    const testRenderer = renderButtons(onIncrement, onDecrement, 1);
    const testInstance = testRenderer.root;

    // find a single descendant test instance with the provided type
    // if there is not exactly one test instance with the provided type, it will throw an error
    expect(testInstance.findByType(TouchableItem).props.onPress).toBe(
      onIncrement,
    );
  });

  it('should not increment over max', () => {
    const onIncrement = jest.fn();
    const onDecrement = jest.fn();
    const testRenderer = renderButtons(onIncrement, onDecrement, 3);
    const testInstance = testRenderer.root;

    // find a single descendant test instance with the provided type
    // if there is not exactly one test instance with the provided type, it will throw an error
    expect(testInstance.findByType(TouchableItem).props.onPress).toBe(
      onDecrement,
    );
  });
});
