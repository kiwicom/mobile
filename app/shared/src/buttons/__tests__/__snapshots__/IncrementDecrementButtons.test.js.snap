// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncrementDecrementButtons renders 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderColor": "#0097a9",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "height": 29,
      "width": 94,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#0097a9",
        "borderRightWidth": 1,
        "paddingBottom": 0,
      }
    }
  >
    <Button
      onPress={[MockFunction]}
      style={
        Object {
          "paddingBottom": 32,
        }
      }
      text="â€“"
      touchable={false}
    />
  </Component>
  <Button
    onPress={[MockFunction]}
    text="+"
    touchable={true}
  />
</Component>
`;
